### PR TITLE
Support for 2-stage loading via shim and systemd-boot

### PR DESCRIPTION
**Pull request by busykai**
This introduces 2-staged boot scheme using `shim` as the first stage loader and `systemd-boot` as the second. New scheme assumes a new ESP layout which is defined as follows:

```
        /EFI/
            Boot/
                BOOTX64.EFI         <-- whatever was there before

            org.clearlinux/
                bootloaderx64.efi       <-- shim
                loaderx64.efi           <-- systemd-boot bootloader
                mmx64.efi               <-- MOK mgr
                fbx64.efi                  <-- Fallback bootloader

            kernel/                 <-- kernels
                kernel-org.clearlinux.....
                ...

        /loader/                    <-- systemd-boot config
            entries/                <-- boot menu entries
                Clear-linux...conf
                ...
            loader.conf             <-- bootloader config

```

One of the distinctive features of this scheme is that it does not modify the default fallback loader, but uses EFI Boot variables instead (see `bootvar.[ch]`).

Decision with regards to clean up mechanism (removing kernels, directories, config entries from previous layouts/boot schemes): this PR will not implement any. Since the new scheme implemented here is non-intrusive and leaves intact previous boot paths, it is not necessary to implement the clean up right now (especially since it's not something readily provided by clr-boot-manager). We will think of a mechanism to perform cleanups later on. Leaving artifacts on ESP is not a nice thing, but it does not impact the functionality.

TODO list:
- [x] ~~Clean up after previous versions of systemd-boot~~
- [x] Fix/develop tests.
- [x] Revisit/improve error handling for the new scheme
- [x] Revisit/improve diagnostics output for the new scheme
- [x] Add command-line support for avoiding creation of EFI var (needed for testing)

CC: @bryteise, @ahkok

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -40,8 +42,6 @@ script:
     - lcov --version | grep "1.10"
     - export CC="gcc-5"
     - ./configure --enable-coverage && make && make check && make distcheck && make check-valgrind
-    - clang-format-3.8 -i $(find . -name '*.[ch]') && git diff --exit-code
```
Comment (ahkok) [inline .travis.yml]: if you have to remove this, you need to do so for a good reason, and state exactly in what case it is broken and why that shouldn't be fixed in clang-format instead upstream.

Disabling this without further comment just means you've got style issues and you don't want to fix them, which is OK for WIP patches but they should get fixed nonetheless :)

File: Makefile.am [hunk 2]
```diff
@@ -42,7 +42,8 @@ AM_CFLAGS = -fstack-protector -Wall -pedantic \
         -Wall -W -D_FORTIFY_SOURCE=2 \
         -Wno-missing-field-initializers \
         -std=c11 \
-        -DSYSCONFDIR=\"$(sysconfdir)\"
+        -DSYSCONFDIR=\"$(sysconfdir)\" \
+	    -D_POSIX_C_SOURCE=201112L
```
Comment (ahkok) [inline Makefile.am]: tabs/spaces?

File: src/bootloaders/shim-systemd.c [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        .needs_install = shim_systemd_needs_install,
+        .needs_update = shim_systemd_needs_update,
+        .install = shim_systemd_install,
+        .update = shim_systemd_update,
+        .remove = shim_systemd_remove,
+        .destroy = shim_systemd_destroy,
+        .get_capabilities = shim_systemd_get_capabilities
+};
+
+#if UINTPTR_MAX == 0xffffffffffffffff
```
Comment (ahkok) [inline src/bootloaders/shim-systemd.c]: maybe this should be something that `configure` picks up? seems critical to know this up front.

File: src/bootloaders/shim-systemd.c [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+static bool shim_systemd_set_default_kernel(const BootManager *manager, const Kernel *kernel) {
+        /* this writes systemd config. systemd has the configuration paths
+         * hardcoded, hence whatever sd_class is doing is OK. */
+        return sd_class_set_default_kernel_impl(manager, kernel, NULL);
+}
+
+static bool exists_identical(const char *path, const char *spath) {
+        if (!nc_file_exists(path)) return false;
```
Comment (ahkok) [inline src/bootloaders/shim-systemd.c]: style comment: No matter what, after the condition I want to see a newline. Either with or without braces (I think this project uses braces after every `if`) but newline required.

Comment (ahkok) [reply src/bootloaders/shim-systemd.c]: Obviously this comment applies to all the occurrences below as well.

File: tests/harness.c [hunk 5]
```diff
@@ -322,6 +332,23 @@ bool push_bootloader_update(int revision)
                 fprintf(stderr, "Failed to update bootloader: %s\n", strerror(errno));
                 return false;
         }
+#if defined(HAVE_SHIM_SYSTEMD_BOOT)
+        {
```
Comment (ahkok) [inline tests/harness.c]: bracket needed here? seems cosmetic afaics

**Comment by ahkok** [comment]
There's 3 open boxes in the commit, I assume they still need addressing?

**Comment by ahkok** [comment]
FWIW it mostly looks good other than style issues. I think we just want to get thorough testing on this done, somehow?

File: src/bootloaders/shim-systemd.c (hunk 3 above)
Comment (busykai) [reply src/bootloaders/shim-systemd.c]: Yeah, there's like 25 places this is done. 26th with this one. `config.h` should really have them. I'm not sure this PR should be the place to fix this though.

File: .travis.yml (hunk 1 above)
Comment (busykai) [reply .travis.yml]: Actually I did this on purpose to have this discussion. clang-format does ugly things like [this](https://github.com/clearlinux/clr-boot-manager/pull/75/files#diff-bd3215aa838aba5c7f9c9b68fe10caddL539).

I do want to question use of automated formatting tool with 8-space indent, ~20 char long identifiers, and line width of 80. And no, I don't think I have issues with style. I do take an issue with what clang-format [wants me to do](https://travis-ci.org/clearlinux/clr-boot-manager/builds/271098489#L2057).

File: src/bootloaders/shim-systemd.c (hunk 4 above)
Comment (busykai) [reply src/bootloaders/shim-systemd.c]: > No matter what, after the condition I want to see a newline.
[C substance](https://github.com/btrask/stronglink/blob/master/SUBSTANCE.md) (which is recommended guideline) and I disagree with you. :)

File: .travis.yml (hunk 1 above)
Comment (bryteise) [reply .travis.yml]: clang-format will do ugly things at times. I think it is reasonable to consider moving the 80 character line width and 8 space indent. At this point it is just a simple matter of tweaking some configuration files and rerunning clang-format once we have a consensus so the bike shed is the painful part.

How about 4 space indent and 100 character line width? (I am not really concerned on the indent but don't want line width to stretch too far, for my personal preferences)

File: src/bootloaders/shim-systemd.c [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        snprintf(path, PATH_MAX, "%s%s", boot_root, KERNEL_DST_DIR);
+        if (!nc_mkdir_p(path, 00755)) return false;
+        snprintf(path, PATH_MAX, "%s%s", boot_root, SYSTEMD_ENTRIES);
+        if (!nc_mkdir_p(path, 00755)) return false;
+        return true;
+}
+
+static bool shim_systemd_install(const BootManager *manager) {
+        char varname[9];
+        (void)manager;
```
Comment (busykai) [inline src/bootloaders/shim-systemd.c]: remove.

File: src/bootloaders/shim-systemd.c (hunk 4 above)
Comment (bryteise) [reply src/bootloaders/shim-systemd.c]: Hrm I disagree a little with that guide (which actually has nice things in it), I like to have different lines for the condition and the conditional statement because it prevents me assuming that I changed both like I thought when I see the one line diff.

Comment (ahkok) [reply src/bootloaders/shim-systemd.c]: It's much easier to observe patterns when split out over 2 lines, the style guide is plainly wrong here IMHO.

File: .travis.yml (hunk 1 above)
Comment (busykai) [reply .travis.yml]: Well, yes, it's a bikeshed discussion. I simply re-enabled it and made it happy for the sake of this PR. The code in general have a formatting/style issue IMHO, but it should be addressed later by an inspired volunteer (like me on a weekend).

Having said that I'm fine with either 80 or 100 char width. 100 is OK even on a vga console. I'd suggest that we use 4-space indent. Kernel-style indenting does not look good when long identifiers are used for variables and functions.

If we agree on this, I'll file an issue and fix it in its own PR.

File: src/bootloaders/shim-systemd.c (hunk 4 above)
Comment (busykai) [reply src/bootloaders/shim-systemd.c]: OK. New line with braces then. No braces are only good for 1-liners...

File: Makefile.am (hunk 2 above)
Comment (busykai) [inline Makefile.am]: Yes. Needs fixing.

File: tests/harness.c (hunk 5 above)
Comment (busykai) [reply tests/harness.c]: it's just to have scope for an isolated task and its variables declared where they belong (both in scope and the beginning of it). is it really an issue?

File: Makefile.am (hunk 2 above)
Comment (busykai) [reply Makefile.am]: @ahkok, so i made `Makefile.am` use spaces and tabs consistently throughout and git decided it had to be re-written. It's slightly more pleasant [to view](https://github.com/busykai/clr-boot-manager/blob/kai/shim-systemd/Makefile.am) now.

**Comment by busykai** [comment]
@ikeydoherty, are you interested in taking a quick look at this?

File: src/bootloaders/bootloader.h [hunk 7]
```diff
@@ -21,6 +21,7 @@
 
 typedef bool (*boot_loader_init)(const BootManager *);
 typedef bool (*boot_loader_install_kernel)(const BootManager *, const Kernel *);
+typedef char *(*boot_loader_get_kernel_dst)(const BootManager *);
```
Comment (ikeydoherty) [inline src/bootloaders/bootloader.h]: Any reason for short-handing the names?

File: src/bootloaders/shim-systemd.c [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+                               .get_capabilities = shim_systemd_get_capabilities };
+
+#if UINTPTR_MAX == 0xffffffffffffffff
+#define EFI_SUFFIX "x64.efi"
+#else
+#define EFI_SUFFIX "ia32.efi"
+#endif
+
+/* Layout entries, see the layout description at the top of the file. */
+#define SHIM_SRC_DIR "usr/lib/shim"
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: I feel these defines could be cleaner to read and better named, its kinda hard to follow whats meant to do what, and its not configurable

File: src/bootloaders/shim-systemd.c [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+        if (!nc_file_exists(path))
+                return false;
+        if (spath && !cbm_files_match(path, spath))
+                return false;
+        return true;
+}
+
+static bool shim_systemd_needs_install(const BootManager *manager)
+{
+        (void)manager;
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: we have `__cbm_unused__` to mark a parameter as unused, no need for the `(void)` cast (which doesn't always work with older compilers)

File: src/bootloaders/shim-systemd.c [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+                return true;
+        if (!exists_identical(systemd_dst_host, systemd_src))
+                return true;
+        return false;
+}
+
+static bool make_layout(const BootManager *manager)
+{
+        char *boot_root = boot_manager_get_boot_dir((BootManager *)manager);
+        char path[PATH_MAX];
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: Any reason we're moving from our safe asprintf (string_printf) helpers to snprintf usage?

File: src/bootloaders/shim-systemd.c [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+        if (!copy_file_atomic(shim_src, shim_dst_host, 00644)) {
+                LOG_FATAL("Cannot copy %s to %s", shim_src, shim_dst_host);
+                return false;
+        }
+        if (!copy_file_atomic(systemd_src, systemd_dst_host, 00644)) {
+                LOG_FATAL("Cannot copy %s to %s", systemd_src, systemd_dst_host);
+                return false;
+        }
+
+        if (bootvar_create(BOOT_DIRECTORY, shim_dst_esp, varname, 9)) {
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: It might be more pertinent to make this ERROR not FATAL, as dodgy firmware implementations are just going to fail here if there is no room left for new vars

File: src/bootloaders/shim-systemd.c [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+        /* init systemd-class since we're reusing it for kernel install.
+         * specific values do not matter as long as sd_class is not used to
+         * install the bootloaders themselves. */
+        static BootLoaderConfig systemd_config = {.vendor_dir = "systemd",
+                                                  .efi_dir = "/usr/lib/systemd/boot/efi",
+                                                  .efi_blob = "systemd-boot" EFI_SUFFIX,
+                                                  .name = "systemd-boot" };
+        sd_class_init(manager, &systemd_config);
+
+        prefix = strdup(boot_manager_get_prefix((BootManager *)manager));
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: We have `nc_string` APIs for dealing with strings and lengths, which ensure safe string operations, and track the length for you through (ab)use of asprintf

File: src/bootloaders/shim-systemd.c [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+        len = strlen(prefix);
+        if (len > 0 && prefix[len - 1] == '/')
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: Shouldn't you be using `basename` or `parent` APIs from Nica instead of mangling the strings? :)

File: src/bootloaders/shim-systemd.c [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+
+        free(prefix);
+        free(boot_root);
+
+        return true;
+}
+
+static void shim_systemd_destroy(const BootManager *manager)
+{
+        (void)manager;
```
Comment (ikeydoherty) [inline src/bootloaders/shim-systemd.c]: more `__cbm_unused__` stuffs

File: src/bootloaders/systemd-class.c [hunk 15]
```diff
@@ -174,7 +177,15 @@ static bool sd_class_ensure_dirs(__cbm_unused__ const BootManager *manager)
...
-bool sd_class_install_kernel(const BootManager *manager, const Kernel *kernel)
+char *sd_class_get_kernel_dst(const BootManager *manager)
+{
+        (void)manager;
+        return strdup(sd_class_config.kernel_dir);
+}
+
+bool sd_class_install_kernel_impl(const BootManager *manager, const Kernel *kernel,
+                                  char *(*get_kernel_dst)(const BootManager *),
+                                  bool (*ensure_layout)(const BootManager *))
```
Comment (ikeydoherty) [inline src/bootloaders/systemd-class.c]: Seems odd to pass function pointers around like this, and not the actual bootloader struct or similar. Makes the API look kinda janky.

Comment (ikeydoherty) [reply src/bootloaders/systemd-class.c]: Alternatively having sane typedefs for the function pointers would clean it up.

File: src/bootman/kernel.c [hunk 16]
```diff
@@ -592,36 +592,26 @@ bool boot_manager_install_kernel_internal(const BootManager *manager, const Kern
         autofree(char) *base_path = NULL;
         autofree(char) *initrd_target = NULL;
         const char *initrd_source = NULL;
-        bool is_uefi = false;
-        autofree(char) *efi_boot_dir = NULL;
+        int is_uefi = (manager->bootloader->get_capabilities(manager) & BOOTLOADER_CAP_UEFI);
```
Comment (ikeydoherty) [inline src/bootman/kernel.c]: Why int? Should be a `bool` to make it clear that we're set to true if the capabilites export `BOOTLOADER_CAP_UEFI`

File: src/bootman/kernel.c [hunk 17]
```diff
@@ -592,36 +592,26 @@ bool boot_manager_install_kernel_internal(const BootManager *manager, const Kern
...
-                                  strerror(errno));
-                        return false;
-                }
-                kfile_target = string_printf("%s/%s", efi_boot_dir, kernel->target.path);
-        } else {
-                kfile_target = string_printf("%s/%s", base_path, kernel->target.legacy_path);
-        }
+        /* for UEFI, the kernel location is prefixed with efi_boot_dir which is
+         * guaranteed to start with '/' since it's its absolute path on ESP. */
+        kfile_target = string_printf("%s%s/%s",
```
Comment (ikeydoherty) [inline src/bootman/kernel.c]: This is less clean/easy to follow than before.

File: src/bootman/kernel.c [hunk 18]
```diff
@@ -676,33 +665,31 @@ bool boot_manager_remove_kernel_internal(const BootManager *manager, const Kerne
...
-        }
+        kfile_target = string_printf("%s%s/%s",
+                                     base_path,
+                                     (is_uefi ? efi_boot_dir : ""),
+                                     (is_uefi ? kernel->target.path : kernel->target.legacy_path));
 
         if (kernel->source.initrd_file) {
-                initrd_target = string_printf("%s/%s",
-                                              is_uefi ? efi_boot_dir : base_path,
+                initrd_target = string_printf("%s%s/%s",
```
Comment (ikeydoherty) [inline src/bootman/kernel.c]: like before - this is less clean

File: src/bootman/kernel.c (hunk 16 above)
Comment (busykai) [reply src/bootman/kernel.c]: with `int` the expression is shorter (no need to compare with the value of the bit flag).

File: src/bootman/kernel.c (hunk 17 above)
Comment (busykai) [reply src/bootman/kernel.c]: the comment explains the why, does it not?

File: src/bootloaders/bootloader.h (hunk 7 above)
Comment (busykai) [reply src/bootloaders/bootloader.h]: you mean 'dst' for 'destination'?

File: src/bootloaders/shim-systemd.c (hunk 10 above)
Comment (busykai) [reply src/bootloaders/shim-systemd.c]: the question is rather why use heap for local variables (and 3 times so)?

Comment (bryteise) [reply src/bootloaders/shim-systemd.c]: So in the happy case of nothing weird went wrong with the boot_root or whatever causing it to be > PATH_MAX this code is more performant but in the unhappy case where boot_root somehow (I'm yet to track that down) is too big for PATH_MAX this snprintf will trim off characters making the path actually wrong but the mkdir may not fail which would be really nasty. I'd just want a fail safe that boot_root's size + any of the DST_DIR, KERNEL_DST_DIR and SYSTEMD_ENTRIES is under the PATH_MAX size. asprintf avoids these checks since the mkdir will just fail.

File: src/bootloaders/systemd-class.c (hunk 15 above)
Comment (busykai) [reply src/bootloaders/systemd-class.c]: I, personally, do not see anything clunky about passing a pointer to a function (because it's C). The logic requires this so-called class to call different implementations depending on the context. I see the problem with the design (which I agree exists) differently: to me the root cause of these pirouettes is the fact that `BootManager` is completely opaque. The very best solution to this is to expose the structure of `BootManager`, at least partially, to be able to access (read-only) the context information and get rid of these completely unnecessary `_impl` functions. However, the purpose of this PR is to make a functional implementation without burying in design discussions.

I'll make the change to pass `BootLoader` to `_impl` functions, but, again, it only masks the design issue (so would `typedef`ing the functions).

File: src/bootloaders/shim-systemd.c (hunk 11 above)
Comment (busykai) [reply src/bootloaders/shim-systemd.c]: if the EFI var cannot be created, then it's a fatal error for this boot scheme.

File: src/bootloaders/shim-systemd.c (hunk 10 above)
Comment (busykai) [reply src/bootloaders/shim-systemd.c]: i do not agree with this at all. it's a static aux function which uses a trusted callee. even if a call to this implementation may result in undesired side effects, the root cause can never be this implementation itself. but i will fix the code to support important use cases when `strlen(boot_var)` is comparable to addressable process memory. :)

**Comment by busykai** [comment]
@ahkok, @bryteise, @ikeydoherty this is ready for final review. As the result of testing, I've added installation of fallback (default) bootloader in image mode so that the resulting image would boot without EFI vars. I'm continuing with testing for another couple of days and will update the PR and let you know if anything else pops up.

File: src/lib/bootvar.c [hunk 19]
```diff
@@ -38,6 +38,11 @@
  * enough. actual space occupied is normally >2 times less. */
 #define BOOT_VAR_MAX 1024
 
+/* env var to turn on the test mode. if set to "yes", turns off any side-effects
+ * functions defined in this file may have: the call will always succeed without
+ * side effects. */
+#define CBM_BOOTVAR_TEST_MODE_VAR "CBM_BOOTVAR_TEST_MODE"
```
Comment (bryteise) [inline src/lib/bootvar.c]: Could we eventually get a vtable setup so mocking bootvar_* functions can be inserted in tests so we can add code coverage for a few of the functions that don't touch hardware? (Future todo perhaps).

File: src/bootman/kernel.c (hunk 16 above)
Comment (bryteise) [reply src/bootman/kernel.c]: @busykai I may be a little confused are you saying eventually it may be useful to keep the is_uefi around as a set of bit flags (hense int would make more sense)? Otherwise I agree with @ikeydoherty and think the intended meaning is clearer with bool (the variable name indicates true or false to me too).

File: src/bootloaders/shim-systemd.c [hunk 20]
```diff
@@ -196,12 +205,33 @@ static bool make_layout(const BootManager *manager)
         if (!nc_mkdir_p(path, 00755)) {
                 goto fail;
         }
+        /* in case of image creation, override the fallback bootloader, so the
+         * media will be bootable. */
+        if (boot_manager_is_image_mode((BootManager *)manager)) {
+                snprintf(path, PATH_MAX, "%s%s", boot_root, EFI_FALLBACK_DIR);
```
Comment (bryteise) [inline src/bootloaders/shim-systemd.c]: This should switch to use the asprintf based functions too.

**Comment by bryteise** [comment]
Okay, I have added my final comments. @ikeydoherty @ahkok please add your final comments or approval today.

**Comment by ikeydoherty** [comment]
@bryteise per IRC, the codestyle is even more out of whack with the existing codebase than the last time I looked. Currently the codebase requires curl braces *even for* single line statements - to ensure consistency (and its frowned on to not do so.)

Plus the truncated word style is still there (`_dst` in function names, etc)

I brought similar points up last time which were contested, and I feel its degraded since. That's my honest response.

**Comment by bryteise** [comment]
@ikeydoherty I thought the _dst function rename hit (or I remember seeing the commit). It is a bit hard to follow in the history though =/.

I think I missed the curly brace style change though.

**Comment by ikeydoherty** [comment]
Apologies github is forcibly caching views, the `dst` thing is fixed. Curly braces thing still stands though :]

File: src/bootman/kernel.c (hunk 16 above)
Comment (busykai) [reply src/bootman/kernel.c]: i'll just change this to `bool`.
